### PR TITLE
gh-108590: Improve sqlite3 docs on encoding issues and how to handle those

### PR DESCRIPTION
**Pull request by erlend-aasland**
Document how to handle table columns with invalid Unicode sequences.

* Issue: gh-108590

----
:books: Documentation preview :books:: https://cpython-previews--108699.org.readthedocs.build/

### CODE REVIEW

**Comment by erlend-aasland** [comment]
IMO, the existing `text_factory` examples looks contrived; the added example where we solve the issue of invalid Unicode sequences is a more interesting and useful example.

**Comment by erlend-aasland** [comment]
cc. @AlexWaygood, if you have time to look at the prose.

File: Doc/library/sqlite3.rst [hunk 1]
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
...
+How to create and use text factories
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for correctly encoded UTF-8 text, but it will fail for invalid
+Unicode sequences and other encodings.
+To work around this, you can use a custom :attr:`~Connection.text_factory`.
+
+Because of SQLites flexible typing, it is not uncommon to encounter table
```
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]:
```suggestion
Because of SQLite's flexible typing, it is not uncommon to encounter table
```

File: Doc/library/sqlite3.rst [hunk 2]
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
...
+columns with the ``TEXT`` data type, containing arbitrary data.
+Let's create a test database with an invalid Unicode sequence:
```
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]: Maybe make clear that this isn't necessarily a _recommendation_, just a demonstration:

```suggestion
To demonstrate, let's create a test database with an invalid Unicode sequence:
```

Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]: Maybe make clear that this isn't necessarily a _recommendation_, just a demonstration:

```suggestion
To demonstrate, let's create a test database with an invalid Unicode sequence:
```

File: Doc/library/sqlite3.rst [hunk 3]
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
...
+
+.. testcode::
+
+   con = sqlite3.connect(":memory:")
+   con.executescript("""
+       CREATE TABLE test (data TEXT);
+       INSERT INTO test VALUES(CAST(X'619F' AS TEXT));
+   """)
+
+To work with such databases, we can use the following trick,
```
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]: "Trick" sounds a _tad_ colloquial here ;)

```suggestion
To work with such databases, we can use the following technique,
```

File: Doc/library/sqlite3.rst [hunk 4]
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
...
+The dump will now print with Unicode surrogate escapes:
+
+.. testoutput::
+
+   BEGIN TRANSACTION;
+   CREATE TABLE test (data TEXT);
+   INSERT INTO "test" VALUES('a\udc9f');
+   COMMIT;
+
+Notice that in order to write the invalid Unicode sequence to a file,
```
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]:
```suggestion
Note that in order to write the invalid Unicode sequence to a file,
```

File: Doc/library/sqlite3.rst [hunk 5]
```diff
@@ -2614,6 +2583,65 @@ With some adjustments, the above recipe can be adapted to use a
...
+you must also use ``errors="surrogateescape"`` as an argument to :func:`open`:
```
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]:
```suggestion
you will also have to use ``errors="surrogateescape"`` as an argument to :func:`open`:
```

**Comment by erlend-aasland** [comment]
Thanks, Alex!

**Review by serhiy-storchaka — COMMENTED** [review]
`print()` will fail, because `sys.stdout.errors` is "strict" by default. It is not a new issue, you have the same issue when you print the result of `os.listdir()`, for example.

Other issue is that you cannot simply feed the result of `iterdump()` to `execute()`, because the Python interface only accept UTF-8 encodable string.

A minor problem is that using anything besides `str`, `bytes` or `bytearray` as `text_factory` slows down requests. So you only need to use this if necessary.

File: Doc/library/sqlite3.rst [hunk 6]
```diff
@@ -2614,6 +2590,70 @@ With some adjustments, the above recipe can be adapted to use a
...
+We will use a `CAST expression`_ to coerce an invalid Unicode sequence,
+represented as a hexadecimal string ``X'619F'``,
+into the ``TEXT`` data type:
+
+.. testcode::
+
+   con = sqlite3.connect(":memory:")
+   con.executescript("""
+       CREATE TABLE test (data TEXT);
+       INSERT INTO test VALUES(CAST(X'619F' AS TEXT));
```
Comment (serhiy-storchaka) [inline Doc/library/sqlite3.rst]: Are you from Norway? Did you use some historical character encodings different from ISO-8859-1?

I think that you can use more real-world example here. Some Norwegian geographic or personal name containing non-ASCII characters encoded in historical Norwegian character encoding. It is better to use something different from Latin1 and cp1252 because they are used sometimes as default fallback encoding, and we want to show that it is not the case here.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: > Are you from Norway?

I am.

> Did you use some historical character encodings different from ISO-8859-1?

Not that I can recall. IIRC, most webpages back in the 90s used ISO-8859-1.

> I think that you can use more real-world example here.

I agree; real-world examples are always better than contrived examples.

> Some Norwegian geographic or personal name containing non-ASCII characters encoded in historical Norwegian character encoding. It is better to use something different from Latin1 and cp1252 because they are used sometimes as default fallback encoding, and we want to show that it is not the case here.

Let's try and find something :)

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: cc. @corona10 (as mentioned on the Discord chat)

**Comment by CorvinM** [comment]
It may be helpful to emphasize that the surrogate escapes cannot be passed back to sqlite through python.

**Comment by serhiy-storchaka** [comment]
The plausible real-world example -- a DB created from a CSV file or other source in legacy encoding.

Ukrainian or Korean are bad as examples, because _all_ characters are non-ASCII, so you cannot not recognize the word in its bytes repr. "Österreich" is better, it contains only one non-Latin letter. Of course you need to use the encoding different from UTF-8 for example. It would be better if the natural legacy encoding is different from Latin1, but if there is nothing better, Latin1 works too. "Österreich" is a bit too long, some shorter Norwegian geographical name may be better.

The new example only shows the use of `iterdump()` and does not show the use of `bytes`. I think that examples with `fetch()` are important too. You can show it for `bytes`, `encoding='latin1'` and `errors='surrogateescape'`, then say that `iterdump()` only works with `text_factory` producing a string and show how to save it in a file (and you can chose whether you want to recreate the DB with the originally encoded data or recode it in UTF-8 if you know the used encoding).

**Comment by encukou** [comment]
The CSV export from my Czech bank account uses iso-8859-2 and contains the column `b'\xc8\xe1stka'`.
*Částka* means "sum" (quantity of money). Decoding with Latin-1 this gives you `Èástka` which is wrong.

Or you could use `b'Polo\xbeka'` (*položka*, "entry"), which latin1 decodes to the more obviously wrong `Polo¾ka`.

**Comment by serhiy-storchaka** [comment]
Good examples!

File: Doc/library/sqlite3.rst [hunk 7]
```diff
@@ -1157,6 +1157,13 @@ Connection objects
                  f.write('%s\n' % line)
          con.close()
 
+      .. note::
+
+         If your database contains ``TEXT`` values with invalid Unicode
+         sequences, or encodings incompatible with UTF-8,
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: What does "invalid Unicode sequences" refer to?  The database should contain bytes, and it seems that these bytes are expected to be valid UTF-8.  There are bytes sequences that are invalid in UTF-8 (but might be valid for other encodings), but if the UTF-8 decoding succeeds, then the resulting sequence of Unicode characters should be ok (unless we are talking about surrogate pairs, some combining characters, or something else that I might be missing).

File: Doc/library/sqlite3.rst [hunk 8]
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
...
 
+.. _sqlite3-howto-text-factory:
+
+How to create and use text factories
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for correctly encoded UTF-8 text, but it will fail for invalid
+Unicode sequences and other encodings.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: Here it should probably just talk about "valid UTF-8" or "text correctly encoded as UTF-8" (note that "UTF-8 text" doesn't really exist since UTF-8 is bytes, and you can't encode UTF-*).

File: Doc/library/sqlite3.rst [hunk 9]
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
...
+To work around this, you can use a custom :attr:`~Connection.text_factory`.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: Even though defaulting on UTF-8 is technically a limitation, I wouldn't say it's something that needs to be "worked around"
```suggestion
You can use a custom :attr:`~Connection.text_factory` to specify different encodings.
```

File: Doc/library/sqlite3.rst [hunk 10]
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
...
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing arbitrary data.
+To demonstrate, let's create a test database with an invalid Unicode sequence.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: Same thing, you want to create a field that contains a byte sequence that is not valid UTF-8 (or that can't be decoded as UTF-8).

File: Doc/library/sqlite3.rst (hunk 6 above)
Comment (ezio-melotti) [reply Doc/library/sqlite3.rst]: Using a different iso-8859-X (where `X != 1`) might be a good compromise, even though all iso-8859-X encoding can decode any byte sequence without decoding errors (even though the result might be unexpected).

> encoded in historical Norwegian character encoding

I'm not aware of any specific encoding other than `iso-8859-1`.
The Runic Unicode block comes to mind, but afaik that can only be encoded by UTF-* encodings :upside_down_face:), and wouldn't solve the problem at hand.

File: Doc/library/sqlite3.rst [hunk 11]
```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
...
+borrowed from the :ref:`unicode-howto`:
+
+.. testcode::
+
+   con.text_factory = lambda data: str(data, errors="surrogateescape")
+   dump = con.iterdump()
+   for line in dump:
+       print(line)
+
+The loop above will print the offending line using Unicode surrogate escapes:
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: "offending" might be a bit harsh if the DB simply uses a different encoding for all the text.  If instead a non-UTF-8 byte sequence managed to sneak in among the rest of the UTF-8 then the term might ok (even though I would be more offended by the code that would have let that happen in the first place).

Maybe it would be useful to address these two cases separately?

File: Doc/library/sqlite3.rst (hunk 7 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: We're talking about byte sequences that cannot be decoded to valid UTF-8. For example:

```python
>>> b"\x61\x9f".decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x9f in position 1: invalid start byte
```

File: Doc/library/sqlite3.rst (hunk 8 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Good point :) I'll reword it tomorrow.

File: Doc/library/sqlite3.rst (hunk 9 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Yeah, "work around" has a negative touch to it. Perhaps:

```suggestion
You can use a custom :attr:`~Connection.text_factory` to handle such cases.
```

File: Doc/library/sqlite3.rst (hunk 10 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Yes, this is poorly worded. I'll suggest a more accurate wording tomorrow.

File: Doc/library/sqlite3.rst (hunk 11 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: > "offending" might be a bit harsh if the DB simply uses a different encoding for all the text.

+1! 😄 

> Maybe it would be useful to address these two cases separately?

Perhaps we only need one of them; I think the case of a different encoding makes for the better example.

Comment (ezio-melotti) [reply Doc/library/sqlite3.rst]: Maybe this is something that we could revisit in a separate PR, but I have the feeling that the focus of this howto is to explain how to use a certain API (namely creating/using text factories) rather than solving a real world problem (i.e. how to deal with non-UTF-8 text fields).  IOW, I think it's unlikely that users will come to the docs looking for text factories (i.e. a solution they might not be aware of) -- they will probably come with a decoding error that they can't figure out (i.e. a problem).

The decoding error might have happened for different reasons depending on what exactly it's on those fields, and the way you deal with it also changes:
* for fields that simply use a different encoding, the solution is to use a text factory that decodes the text using the correct encoding;
* for fields that should be using UTF-8 but contain some ill-encoded byte sequence, the solution is to use `surrogateescape` and friends to escape/ignore the invalid bytes.

Since both problems are related to encodings and can be solved with text factories, you could cover both in the same howto, by explaining the "different encoding" scenario and then just adding a paragraph at the end to mention how to use `surrogateescape`.

File: Doc/library/sqlite3.rst (hunk 6 above)
Comment (serhiy-storchaka) [reply Doc/library/sqlite3.rst]: @encukou proposed to use a Czech example with iso-8859-2, and I think a Czech example is perfect here.

**Review by serhiy-storchaka — COMMENTED** [review]
I think we all agree that the current version has serious issues. No need to beat it more, it is already dead. Let @erlend-aasland to address our comments and prepare a new version. Then we will start a new round of bikeshedding.

**Comment by erlend-aasland** [comment]
Well, turns out bank exports is not a good example for showing how to handle encoding issues in table column data (not table column names). Instead of bikeshedding the example into pieces, I suggest simply using a Czech-English dictionary as an example. _We can then revisit the example itself in a later PR_. I'll push an update in some hours.

An exercise for the reader: try getting SQLite to create non-UTF-8 column names. I won't include the spoiler.

**Comment by serhiy-storchaka** [comment]
Easy-peasy with the `sqlite` CLI:
```
$ echo 'create table таблиця (поле text);' | iconv -t koi8-u | sqlite3 test.db
$ sqlite3 test.db 'select * from sqlite_master;'
table|�������|�������|2|CREATE TABLE ������� (���� text)
$ sqlite3 test.db 'select * from sqlite_master;' | iconv -f koi8-u
table|таблиця|таблиця|2|CREATE TABLE таблиця (поле text)
```

**Comment by erlend-aasland** [comment]
Sure, but that's for a different kind of how to ;)

File: Doc/library/sqlite3.rst (hunk 6 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I hope the example of 3b9f620aacac66c20344473a218af5d75959736f is sufficient for now. Also note that we can cook up a better example in a later PR.

File: Doc/library/sqlite3.rst (hunk 8 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Addressed in 3b9f620aacac66c20344473a218af5d75959736f

File: Doc/library/sqlite3.rst (hunk 10 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: How does 3b9f620aacac66c20344473a218af5d75959736f look?

File: Doc/library/sqlite3.rst (hunk 11 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I reworded the how-to to focus on the problem (encodings) instead of the solution (text factories). I simplified the text (how-to's are not explanations) and tried to cover both cases. How does 3b9f620aacac66c20344473a218af5d75959736f look, Ezio?

**Comment by erlend-aasland** [comment]
@ezio-melotti, would you mind to take another look?

File: Doc/library/sqlite3.rst [hunk 12]
```diff
@@ -2614,6 +2596,65 @@ With some adjustments, the above recipe can be adapted to use a
...
+.. testcode::
+
+   con.text_factory = lambda data: str(data, errors="surrogateescape")
+
+.. note::
+
+   Strings containing surrogate escapes and must be treated with care.
+   You cannot simply pass them back to SQLite,
+   for example using :meth:`~Cursor.execute`,
+   since the :mod:`!sqlite3` module APIs only accept UTF-8 encoded strings.
```
Comment (serhiy-storchaka) [inline Doc/library/sqlite3.rst]: UTF-8 encodable strings.

**Review by serhiy-storchaka — COMMENTED** [review]
It would be nice to document also `text_factory=bytes` option present in the old example. And `iterdumps()` needs special mentioning, because it does not work with `bytes`.

I do not think that it is needed to describe how to create an SQLite file with non-UTF-8 encoding, especially if it involves a third-party tool. It is not the problem which the user needs to solve, their problem is how to work with such files. Such code can go in tests.

File: Doc/library/sqlite3.rst (hunk 12 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Why do you mean a `str` instance is best described as a "UTF-8 encodable string"? Can you elaborate why "encodable strings" is a better term?

**Comment by erlend-aasland** [comment]
> It would be nice to document also `text_factory=bytes` option present in the old example.

Sure. The contrived example in the existing docs, or did you have a specific use case in mind?

> I do not think that it is needed to describe how to create an SQLite file with non-UTF-8 encoding, especially if it involves a third-party tool. It is not the problem which the user needs to solve, their problem is how to work with such files.

I agree.

File: Doc/library/sqlite3.rst (hunk 12 above)
Comment (serhiy-storchaka) [reply Doc/library/sqlite3.rst]: Because if it is UTF-8 encoded, it is no longer a `str` object. It is a bytes sequence, which can be represented in Python as `bytes`, bytearray`, etc. The sqlite module APIs do no support them.

Maybe "UTF-8 encodable string" does not sound well, I do not know, but "UTF-8 encoded string" is not correct here.

File: Doc/library/sqlite3.rst [hunk 13]
```diff
@@ -1562,6 +1539,8 @@ Cursor objects
 
       Use :meth:`executescript` to execute multiple SQL statements.
 
+      :meth:`!execute` only accepts UTF-8 encoded strings.
```
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: @serhiy-storchaka, do you oppose this change?

File: Doc/library/sqlite3.rst [hunk 14]
```diff
@@ -1610,6 +1589,8 @@ Cursor objects
          Starting with Python 3.14, :exc:`ProgrammingError` will
          be raised instead.
 
+      :meth:`!executemany` only accepts UTF-8 encoded strings.
```
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: @serhiy-storchaka, ditto

File: Doc/library/sqlite3.rst [hunk 15]
```diff
@@ -1635,6 +1616,7 @@ Cursor objects
              COMMIT;
          """)
 
+      :meth:`!executescript` only accepts UTF-8 encoded strings.
```
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: @serhiy-storchaka, ditto

File: Doc/library/sqlite3.rst (hunk 12 above)
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: @serhiy-storchaka

Alt 1:

```suggestion
   Strings containing surrogate escapes and must be treated with care.
   You cannot simply pass them back to SQLite.
```

Alt 2:

```suggestion
   Strings containing surrogate escapes and must be treated with care.
   You cannot simply pass them back to SQLite;
   the :mod:`!sqlite3` module API does not understand
   surrogate escape codes.
```

Other suggestions?

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: See https://github.com/python/cpython/pull/108699#discussion_r1314662069 for suggestions.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Alt 3: Just remove the note.

Comment (serhiy-storchaka) [reply Doc/library/sqlite3.rst]: Better ask @ezio-melotti.

File: Doc/library/sqlite3.rst (hunk 13 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Never mind, I removed them; the PR now only focuses on the how-to.

Comment (serhiy-storchaka) [reply Doc/library/sqlite3.rst]: It is technically incorrect. I am not sure what wording would be better here and whether a special addition is needed at all, because for functions have the same requirement, but it is not documented, and in general you cannot know whether surrogate characters are acceptable until you try.

It is not actually clear to what the note refers. Is it only *sql* or *parameters*? Parameters can contain not only strings, but numbers, bytes and custom data. Strings must be UTF-8 encodable, but bytes can be not only UTF-8 encoded.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: *sql* is documented as being a `str` instance; not a `bytes` instance. There is no need to complicate the docs with encoding details here.

File: Doc/library/sqlite3.rst (hunk 12 above)
Comment (CorvinM) [reply Doc/library/sqlite3.rst]: My 2 cents
Alt 4:
```suggestion
   Strings containing surrogate escapes must be treated with care as
   modules that attempt to encode them require support for surrogate
   escapes to handle them correctly.
   The :mod:`!sqlite3` module API does not support surrogate escapes
   so these strings cannot be passed back to the :mod:`!sqlite3` module API.
```

File: Doc/library/sqlite3.rst [hunk 16]
```diff
@@ -1223,6 +1227,10 @@ Connection objects
 
       .. versionadded:: 3.7
 
+      .. seealso:
```
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]:
```suggestion
      .. seealso::
```

File: Doc/library/sqlite3.rst (hunk 12 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]:
```suggestion
   Strings containing surrogate escapes and must be treated with care.
   For example, you cannot pass them back to SQLite,
   since the :mod:`!sqlite3` module API does not support strings
   containing surrogate escape codes.
```

**Comment by erlend-aasland** [comment]
@serhiy-storchaka, do you want to take another look?

**Comment by erlend-aasland** [comment]
@ezio-melotti, would you like to take another look?

File: Doc/library/sqlite3.rst [hunk 17]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
 
+.. _sqlite3-howto-encoding:
+
+How to handle non-UTF-8 text encodings
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for UTF-8 encoded text, but it will fail for other encodings
+and invalid UTF-8.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: This is understandable and probably ok, but technically you might tweak it to say:
```suggestion
This works well for UTF-8 encoded text, but it might fail for other encodings
and invalid UTF-8.
```
Since e.g. ASCII-encoded text will also work (UTF-8 is ASCII-compatible), or even iso-8859-1-encoded text that contain only ASCII will fail.  IOW, it's not guaranteed that having text encoded with a different encoding will result in a failure.

File: Doc/library/sqlite3.rst [hunk 18]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
+You can use a custom :attr:`~Connection.text_factory` to handle such cases.
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing non-UTF-8 encodings,
+or even arbitrary data.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]:
```suggestion
Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
columns with the ``TEXT`` data type containing non-UTF-8 encodings,
or even arbitrary data.
```
Is this comma needed?

File: Doc/library/sqlite3.rst [hunk 19]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
+To demonstrate, let's assume we've got a database with ISO-8859-2 (Latin 2)
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]:
```suggestion
To demonstrate, let's assume we have a database with ISO-8859-2 (Latin 2)
```

File: Doc/library/sqlite3.rst [hunk 20]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
+For invalid UTF-8 or arbitrary data in stored in ``TEXT`` table columns,
+you can use the following technique, borrowed from the :ref:`unicode-howto`:
+
+.. testcode::
+
+   con.text_factory = lambda data: str(data, errors="surrogateescape")
+
+.. note::
+
+   Strings containing surrogate escapes and must be treated with care.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]:
```suggestion
   Strings containing surrogate escapes must be treated with care.
```
Is that a stray `and` or did you want to add something else?

File: Doc/library/sqlite3.rst [hunk 21]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
+   For example, you cannot pass them back to SQLite,
+   since the :mod:`!sqlite3` module API does not support strings
+   containing surrogate escape codes.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: The wording of this sentence is a bit off.  You can talk about surrogates if you are talking about the characters themselves.  `surrogateescape` is an error handler that converts surrogates to Unicode characters in a different range, and you also have escape sequences (`'\uXXXX'`) that represent surrogates.

Does "pass them back" imply that you can get them out from the DB?

File: Doc/library/sqlite3.rst (hunk 18 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: Yeah, the comma implies the structure is:

> it is not uncommon to encounter table columns ((with the ``TEXT`` data type), (containing non-UTF-8 encodings), (or even arbitrary data)).

But it seems the intended meaning is instead:

> it is not uncommon to encounter table columns with the ``TEXT`` data type ((containing non-UTF-8 encodings), (or even arbitrary data)).

File: Doc/library/sqlite3.rst (hunk 19 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: Also, conventional spelling of "Latin-2" is with a hyphen:

```suggestion
To demonstrate, let's assume we have a database with ISO-8859-2 (Latin-2)
```

File: Doc/library/sqlite3.rst [hunk 22]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
+You can use a custom :attr:`~Connection.text_factory` to handle such cases.
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing non-UTF-8 encodings,
+or even arbitrary data.
+To demonstrate, let's assume we've got a database with ISO-8859-2 (Latin 2)
+encoded text, for example a table of Czech-English dictionary entries.
+Assuming we now have a :class:`Connection` instance :py:data:`!con`
+connected to this database,
+we can decode the Latin 2 encoded text using this :attr:`~Connection.text_factory`:
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
we can decode the Latin-2 encoded text using this :attr:`~Connection.text_factory`:
```

Minor tweak

File: Doc/library/sqlite3.rst [hunk 23]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
 instead of a :class:`~collections.namedtuple`.
 
 
+.. _sqlite3-howto-encoding:
+
+How to handle non-UTF-8 text encodings
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
By default, :mod:`!sqlite3` uses :class:`str` to adapt SQLite values
with the ``TEXT`` data type.
```

I find this phrasing clearer, less ambiguous and easier to follow

**Review by CAM-Gerlach — COMMENTED** [review]
Looks very good now, thanks! Just a few small suggestions.

File: Doc/library/sqlite3.rst (hunk 17 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Thanks, good point.

File: Doc/library/sqlite3.rst (hunk 23 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Thanks, that reads better.

File: Doc/library/sqlite3.rst (hunk 21 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Hm, yes this is certainly a bit off. The point we want to highlight (IMO) is that strings containing surrogate escape sequences will not be accepted by the `sqlite3` APIs. For example, `Cursor.execute` will not accept strings with surrogate escape sequences neither as the *sql* string, nor as a part of the *parameters* sequence.

File: Doc/library/sqlite3.rst [hunk 24]
```diff
@@ -2614,6 +2590,49 @@ With some adjustments, the above recipe can be adapted to use a
...
+.. testcode::
+
+   con.text_factory = lambda data: str(data, errors="surrogateescape")
+
+.. note::
+
+   Strings containing surrogate escapes must be treated with care.
+   For example, you cannot pass them back to SQLite,
+   since the :mod:`!sqlite3` module API does not support strings
+   containing surrogate escape codes.
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]:
```suggestion
   The :mod:`!sqlite3` module API does not support strings
   containing surrogates.
```
I would simply say this then.  This will make the users aware of this limitation without going in too much details, so that they will remove the surrogates and try again.  If there are situations where surrogates are allowed (e.g. can I have them in a db and read them?) then it might be worth to specify that.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: The example just above this text shows how to deal with strings w/surrogates in the DB. I think your proposal works swell. Keep it simple 🪄

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Alternatively, one could remove this warning and mention this limitation explicitly for each API.

**Comment by erlend-aasland** [comment]
I think this is ready to land. Shout out if you disagree. OTOH, we can easily patch up things with a new PR. I'll be merging this tomorrow morning CET.

**Comment by erlend-aasland** [comment]
A big thanks to everyone who helped shape this PR!

**Review by serhiy-storchaka — COMMENTED** [review]
This change removes any mention of `bytes`. It makes the documentation less informative.

It also does not resolve the original issue, in particularly, it does not document that `iterdump()` is not compatible with `text_factory=bytes`. And it does not say what to do with the result of `iterdump()` with custom `text_factory`.

**Comment by erlend-aasland** [comment]
> This change removes any mention of `bytes`. It makes the documentation less informative.

Serhiy, please see https://github.com/python/cpython/pull/108699#issuecomment-1704878549. I do not intend to mention `bytes` unless there is a reason to do it. If you care strongly for it, I suggest you create a follow-up PR with your use-case. IMO, it does not block this PR.

> It also does not resolve the original issue, in particularly, it does not document that `iterdump()` is not compatible with `text_factory=bytes`. And it does not say what to do with the result of `iterdump()` with custom `text_factory`.

It does so implicitly; it helps the user understand how to work around such encoding issues. If you disagree with this approach, please open a competing PR.

**Comment by serhiy-storchaka** [comment]
Okay. I do not fully understand what is the benefit of this change, but I do not oppose it.

**Comment by erlend-aasland** [comment]
> Okay. I do not fully understand what is the benefit of this change, but I do not oppose it.

Why so dismissive? I've asked you twice for a concrete proposal regarding your `bytes` comment, but you seem to ignore this. I do not oppose such a change, I only ask for a use-case for such an example.

**Comment by serhiy-storchaka** [comment]
Sorry, I meant no disrespect. There were so many changes and comments in this PR that I was already lost. After the old examples were removed, there was a lot of discussion about new examples, but they never appeared, so I didn't see much to comment on, as this PR looked far from finished. If you consider it complete, I will perhaps continue the work in a new PR. `text_factory` should be fully documented (and special cases for `bytes` and `bytearray` are specially optimized in the code).

**Comment by erlend-aasland** [comment]
Ok, thanks for your response, Serhiy. It is ok to continue improving the docs in follow-up PRs.

Let's land this, then we can follow up your concerns in a new PR. I appreciate your comments; I just find some of them hard to address.

File: Doc/library/sqlite3.rst (hunk 21 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Let's unblock this PR by removing this note; we can add a note in a follow-up PR.
